ldap/servers/slapd/back-ldbm/db-bdb/bdb_bdbreader_glue.c

```diff
@@ -0,0 +1,491 @@
+                if (rc == DB_SUCCESS) {
+                    if (k1.size == key->size && memcmp(k1.data, key->data, k1.size) == 0) {
+                        rc = store_val(rc, dbc, key, data);
+                    } else {
+                        rc = DB_NOTFOUND;
+                    }
+                }
+            }
+            break;
+        case DB_NEXT_NODUP:
+            rc = DB_SUCCESS;
+            while (rc == DB_SUCCESS) {
+                rc = bdbreader_cur_next(dbc->impl);
+                if (rc == DB_SUCCESS) {
+                    rc = bdbreader_cur_getcurval(dbc->impl, &k1.data, &k1.size, NULL, NULL);
+                }
+                if (k1.size != key->size || memcmp(k1.data, key->data, k1.size) != 0) {
+                    rc = store_val(rc, dbc, key, data);
+                    break;
+                }
+            }
+            break;
+        case DB_NODUPDATA:
+            break;
+        case DB_PREV:
+            break;
+        case 0:
+            /* _get_and_add_parent_rdns set the flags ==> probably a bug
+             *   (according BDB C API)
+             * But lets assume it is a DB_SET
+             */
+        case DB_SET:
+            rc = bdbreader_cur_lookup(dbc->impl, key->data, key->size);
+            rc = store_val(rc, dbc, key, data);
+            break;
+        case DB_SET_RANGE:
+            rc = bdbreader_cur_lookup_ge(dbc->impl, key->data, key->size);
+            rc = store_val(rc, dbc, key, data);
+            break;
+        case DB_SET_RECNO:
+            break;
+    }
+
+    slapi_log_err(SLAPI_LOG_ERR, "bdb_ro", "==> dbc_get(flags=%d:%s) rc=%d:%s\n",
+                  flags, LOGN(flags, flagnames), rc, LOGN(rc, errname));
+    if (key && key->data && key->size) {
+        hexadump("Key", key->data, 0, key->size);
+    }
+    if (data && data->data && data->size) {
+        hexadump("Data", data->data, 0, data->size);
+    }
+    return rc;
+}
+
+int dbc_put(DBC *dbc, DBT *key, DBT *data, u_int32_t flags)
+{
+    return DB_NOTSUPPORTED;
+}
+
+int dbc_count(DBC *dbc, void *countp, u_int32_t flags)
+{
+    return DB_NOTSUPPORTED;
+}
+
+int db_cursor(DB *db, DB_TXN *txnid, DBC **cursorp, u_int32_t flags)
+{
+    DBC *dbc = (void*)slapi_ch_calloc(1, sizeof *dbc);
+
+    *cursorp = dbc;
+    dbc->dbp = db;
+
+    dbc->c_close = dbc_close;
+    dbc->c_del = dbc_del;
+    dbc->c_get = dbc_get;
+    dbc->c_put = dbc_put;
+    dbc->c_count = dbc_count;
+    dbc->impl = bdbreader_cur_open(db->impl);
+
+    return (db->impl == NULL) ? DB_OSERROR : DB_SUCCESS;
+}
+
+int db_del(DB *db, DB_TXN *txnid, DBT *key, u_int32_t flags)
+{
+    return DB_NOTSUPPORTED;
+}
+
+int db_put(DB *db, DB_TXN *txnid, DBT *key, DBT *data, u_int32_t flags)
+{
+    return DB_NOTSUPPORTED;
+}
+
+int db_get(DB *db, DB_TXN *txnid, DBT *key, DBT *data, u_int32_t flags)
+{
+    int rc = bdbreader_cur_lookup(db->cur->impl, key->data, key->size);
+    if (rc == DB_SUCCESS) {
+        rc = dbc_get(db->cur, key, data, flags);
```

Comment:
Shouldn't it call db->cur->c_get than dbc_get ?

Reply:
db->cur->c_get == dbc_get so it does not really matter.
But to respect the layered approach since we are in the glue code that maps bdb API to robdb API,
   bdb_bdbreader_glue.c code should use either internal or librobdb function rather than the bdb API
   (So dbc_get looks good.)